### PR TITLE
PACA-822: Skip FISS claims with known-invalid DCN values.

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[PACA-822](https://jira.cms.gov/browse/PACA-822)

**User Story or Bug Summary:**

RDA API has a known issue that allows invalid FISS claims to be stored in their database and streamed to clients.  PACA should not store these claims in BFD database.

Bad FISS claims are those with DCN values consisting entirely of zeros+spaces or those with DCN values containing the letter X in positions 15-17..

---

### What Does This PR Do?

- Renames `getDepupKeyForMessage` to `getClaimIdForMessage` to make its meaning explicit and ensure it is safe for use in logging.
- Adds `isValidMessage` method to `RdaSink` with FISS implementation that validates the DCN is acceptable.
- Modifies `StandardGrpcRdaSource` to use `isValidMessage` to skip invalid messages.
- Adds javadocs to modified methods that had none.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/FissClaimRdaSink.java [hunk 1]
```diff
@@ -13,10 +13,19 @@
...
 
 /** Implementation of AbstractClaimRdaSink that adds FISS claim specific methods. */
 public class FissClaimRdaSink extends AbstractClaimRdaSink<FissClaimChange, RdaFissClaim> {
+  /**
+   * Some FISS claim records received from the IDR are not real claims. These records can be
+   * recognized by their DCN values. Any FISS claim with DCN that is all zeros or ends with XXX is
+   * not a real claim and should not be written to the database. This regex is used by {@link
+   * #isValidMessage} to recognize these bad claims.
+   */
+  private static final Pattern InvalidDcnRegex = Pattern.compile("(^0+$)|(XXX$)");
```
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/FissClaimRdaSink.java:27]: We need to account for leading and trailing whitespace as well.  For example these are two DCNs from PROD that we'd want to skip over.
- '22239999999999XXX      ' 
- ' 00000000000           '

Comment (ericbuckley) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/FissClaimRdaSink.java:27]: Also when reviewing the database, I just noticed something new.  Its not just that the DCN needs to end in 'XXX', its that the [BSI portion](https://confluence.cms.gov/pages/viewpage.action?pageId=540083018) (positions 15-17) of the DCN is 'XXX'.  For example

- '22230501299999XXX4     '

Sorry for the new spec, I just noticed this myself.

**Review by ericbuckley — APPROVED** [review]

**Review by KyRobbins — APPROVED** [review]

**Review by lsmitchell — APPROVED** [review]
Minor non-blocking comment

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/FissClaimRdaSink.java [hunk 2]
```diff
@@ -13,10 +13,19 @@
...
 
 /** Implementation of AbstractClaimRdaSink that adds FISS claim specific methods. */
 public class FissClaimRdaSink extends AbstractClaimRdaSink<FissClaimChange, RdaFissClaim> {
+  /**
+   * Some FISS claim records received from the IDR are not real claims. These records can be
+   * recognized by their DCN values. Any FISS claim with DCN that is all zeros (and/or spaces) or
+   * has X in positions 15-17 is a claim that has been marked as deleted on FISS and should be
+   * ignored. This regex is used by {@link #isValidMessage} to recognize these bad claims.
+   */
+  private static final Pattern InvalidDcnRegex = Pattern.compile("(^[0 ]+$)|(^.{14}XXX)");
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/FissClaimRdaSink.java:27]: @lsmitchell I understand where you're coming from, but the valid DCN format is not easier to check for than the invalid ones.  In particular, the 3 digit BSI always starts with a state abbreviation then trails in a A, B, R or D.  At least thats to our knowledge, what we've been finding is that the shared systems (FISS/MCS/CWF) often make slight changes to their format and don't document them.  For example, in the Home Health segment (pos 18-21) we've seen codes that are not represented in their documentation, but from what we've been told are in fact valid.

From what we've been told by the FISS architects, these two scenarios (all zeros or 'XXX' in positions 15-17) are the only 2 situations where the cobol code is marking the claim as invalid/deleted.   It wouldn't surprise me if there is a third or fourth scenario, but just like the whitelist format, the documentation isn't up-to-date.

Here is some more info on the format of the DCN.  https://confluence.cms.gov/pages/viewpage.action?pageId=540083018

[the remark this answers is not in the document]
Comment (lsmitchell) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/FissClaimRdaSink.java:27]: Thanks for the info! Undocumented changes to what is valid is a pain, makes sense to just check the couple bad ones we know then.

**Pull request merged**